### PR TITLE
Add Implicit Regularized Gaussian notebook by Jasper

### PR DESCRIPTION
**Pull request by nabriis**
Minor modifications to come to show version and author

### CODE REVIEW

File: 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Implicit priors with regularized Gaussian distributions\n",
+    "\n",
+    "### Jasper Everink (jmev@dtu.dk)\n",
+    "\n",
+    "In this notebook, we demonstrate how to use regularized Gaussian distributions from [[1]](https://iopscience.iop.org/article/10.1088/1361-6420/acf9c5) in CUQIpy.\n"
```
Comment (jakobsj) [inline 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb:11]:
```suggestion
    "In this notebook, we demonstrate how to use so-called implicit priors based on regularized Gaussian distributions from [[1]](https://iopscience.iop.org/article/10.1088/1361-6420/acf9c5) in CUQIpy.\n"
```

File: 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+   "source": [
+    "Constraints and regularization can be added to randomized linear least squares problem by adding a suitable penalization function $R(x)$.\n",
+    "The resulting objective function from the regularized, randomized linear least squares problem can be split into two terms:\n",
+    "\n",
+    "$\n",
+    "\\min_{x \\in \\mathbb{R}^n} \\left\\{ \\underbrace{\\frac{1}{2}\\|Ax-\\hat{b}\\|_\\Sigma^2}_{\\text{Gaussian likelihood}}  + \\underbrace{\\frac{1}{2}\\|Lx-\\hat{c}\\|_\\Lambda^2 + R(x)}_{\\text{Implicit regularized Gaussian prior}}\\right\\}.\n",
+    "$\n",
+    "\n",
+    "The Gaussian likelihood term is the same as in the unregularized problem, however, the prior term now consists of both a randomized linear least squares error term from the original Gaussian prior and the regularization term.\n",
+    "These two prior terms are combined into the `RegularizedGaussian` distribution as follows:"
```
Comment (jakobsj) [inline 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb:178]:
```suggestion
    "These two prior terms are combined into the `RegularizedGaussian` implicit prior as follows:"
```

File: 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+    "y  = Gaussian(A@x, 0.001)\n",
+    "BP = BayesianProblem(y, x).set_data(y=y_data)\n",
+    "BP.UQ(exact=info.exactSolution)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that these prior distributions are only defined implicitely and can only be sampled from after being combined with a likehood. Therefore, one cannot sample just form the implicit regularized Gaussian"
```
Comment (jakobsj) [inline 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb:287]:
```suggestion
    "Note that these implicit priors are - as the name suggests - only defined implicitly and can only be sampled from after being combined with a likehood. Therefore, one cannot sample directly from the regularized Gaussian implicit prior itself, which can be seen by the following code causing an error:"
```

File: 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+   "source": [
+    "x  = RegularizedGaussian(0.5*np.ones(n), 0.1, constraint = \"nonnegativity\")\n",
+    "_ = x.sample()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating an implicit regularized Gaussian distribution\n",
```
Comment (jakobsj) [inline 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb:318]:
```suggestion
    "## Creating a regularized Gaussian implicit prior\n",
```

File: 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 5]
```diff
@@ -0,0 +1,581 @@
...
+    "\n",
+    "There are different ways of construction Implicit Regularized Gaussian distributions:\n",
```
Comment (jakobsj) [inline 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb:320]:
```suggestion
    "There are different ways of constructing Regularized Gaussian implicit priors:\n",
```

File: 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+    "- For certain pre-implemented constraints, one can specify the constraint a simple string like \"nonnegativity\" or \"box\".\n",
+    "\n",
+    "- For general constraints on $C$, one can specify a projector function $\\text{projector}(z) = \\argmin_{x \\in C} \\left\\{\\|x-z\\|_2\\right\\}$.\n",
+    "\n",
+    "- For general regularization functions $R$, one has to specify the proximal operator $\\text{prox}_R(z, s) = \\argmin_{x \\in \\mathbb{R}^n} \\left\\{\\|x-z\\|^2_2 + s\\ R(x)\\right\\}$.\n",
+    "\n",
+    "Note that constraints cannot be supplied using an indicator function, due to the algorithm used for solving the regularized linear least squares problem.\n",
+    "\n",
+    "The following code shows various ways of creating implicit regularized Gaussian distributions."
```
Comment (jakobsj) [inline 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb:329]:
```suggestion
    "The following code shows various ways of creating regularized Gaussian implicit priors."
```

File: 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+    "# l1 (R(x) = gamma*\\|x\\|_1) regularization\n",
+    "gamma = 10\n",
+    "x  = RegularizedGaussian(0.5*np.ones(n), 0.1, proximal   = lambda z, s : ProximalL1(z, gamma*s))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Alongside the `RegularizedGaussian` distribution, there is also the `RegularizedGMRF` distribution, which is initialized in a similar manner, but with the `Gaussian` arguments replaced by arguments of `GMRF`."
```
Comment (jakobsj) [inline 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb:365]:
```suggestion
    "Alongside the `RegularizedGaussian` implicit prior, there is also the `RegularizedGMRF` implicit prior, which is initialized in a similar manner, but with the `Gaussian` arguments replaced by arguments of `GMRF`."
```

File: 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Constraints and regularization can be added to randomized linear least squares problem by adding a suitable penalization function $R(x)$.\n",
+    "The resulting objective function from the regularized, randomized linear least squares problem can be split into two terms:\n",
+    "\n",
+    "$\n",
+    "\\min_{x \\in \\mathbb{R}^n} \\left\\{ \\underbrace{\\frac{1}{2}\\|Ax-\\hat{b}\\|_\\Sigma^2}_{\\text{Gaussian likelihood}}  + \\underbrace{\\frac{1}{2}\\|Lx-\\hat{c}\\|_\\Lambda^2 + R(x)}_{\\text{Implicit regularized Gaussian prior}}\\right\\}.\n",
```
Comment (jakobsj) [inline 006_Implicit_Regularized_Gaussian/implicit_priors.ipynb:174]:
```suggestion
    "\\min_{x \\in \\mathbb{R}^n} \\left\\{ \\underbrace{\\frac{1}{2}\\|Ax-\\hat{b}\\|_\\Sigma^2}_{\\text{Gaussian likelihood}}  + \\underbrace{\\frac{1}{2}\\|Lx-\\hat{c}\\|_\\Lambda^2 + R(x)}_{\\text{Regularized Gaussian implicit prior}}\\right\\}.\n",
```

**Review by jakobsj — CHANGES_REQUESTED** [review]
THank you @jeverink for updating to reflect the changes, it is looking great.  I have gone through and only found a few places to replace "distribution" by "implicit prior" and I have also sought to be consistent in using always the term "Regularized Gaussian implicit prior", keeping "implicit" and "prior" together to emphasize the concept "implicit prior".

**Review by jakobsj — APPROVED** [review]
Looks good, thanks!

File: 005_Implicit_Regularized_Gaussian/implicit_priors.ipynb [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Implicit priors with regularized Gaussian distributions\n",
+    "\n",
+    "### Jasper Everink (jmev@dtu.dk)\n",
+    "Regularization is frequently employed in variational inverse problems to enforce constraints on the problem's parameters. An example of such a constraint is sparsity, often induced through regularization techniques like lasso and total variation. In this notebook, we demonstrate how to use so-called implicit priors based on regularized Gaussian distributions from [[1]](https://iopscience.iop.org/article/10.1088/1361-6420/acf9c5) in CUQIpy.\n"
```
Comment (jeverink) [inline 005_Implicit_Regularized_Gaussian/implicit_priors.ipynb:36]: Constraints have a specific meaning in this context and is not the same as assumptions. Perhaps use beliefs instead.

Comment (chaozg) [reply 005_Implicit_Regularized_Gaussian/implicit_priors.ipynb]: Thanks @jeverink , I just replaced constraints with beliefs

**Review by jeverink — APPROVED** [review]

**Review by jakobsj — APPROVED** [review]
Looks great! Thanks for checking and further improving this!

**Comment by nabriis** [comment]
Great. We good to merge?

**Review by chaozg — APPROVED** [review]
LGTM

**Pull request merged**